cmd/nvidia-dra-plugin/device_state.go

```diff
@@ -349,35 +357,70 @@ func (s *DeviceState) prepareDevices(ctx context.Context, claim *resourceapi.Res
 	for _, group := range devices {
-		// Stop any MPS control daemons started for each group of prepared devices.
-		mpsControlDaemon := s.mpsManager.NewMpsControlDaemon(claimUID, group)
-		if err := mpsControlDaemon.Stop(ctx); err != nil {
-			return fmt.Errorf("error stopping MPS control daemon: %w", err)
+		var err error
+		if group.ConfigState.GpuConfig != nil {
+			err = s.unprepareGpus(ctx, group.ConfigState.GpuConfig, group.Devices.Gpus())
 		}
-
-		// Go back to default time-slicing for all full GPUs.
-		tsc := configapi.DefaultGpuConfig().Sharing.TimeSlicingConfig
-		if err := s.tsManager.SetTimeSlice(group.Devices.Gpus(), tsc); err != nil {
-			return fmt.Errorf("error setting timeslice for devices: %w", err)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (s *DeviceState) unprepareGpus(ctx context.Context, config *configapi.GpuConfig, devices PreparedDeviceList) error {
+	if config.DriverConfig.Driver == configapi.VfioPciDriver {
+		for _, device := range devices {
+			if err := s.vfioPciManager.Unconfigure(device.Gpu.Info); err != nil {
+				return fmt.Errorf("error unconfiguring vfio-pci device: %w", err)
+			}
 		}
 	}
+	// Go back to default time-slicing for all full GPUs.
+	tsc := configapi.DefaultGpuConfig().Sharing.TimeSlicingConfig
+	if err := s.tsManager.SetTimeSlice(devices, tsc); err != nil {
+		return fmt.Errorf("error setting timeslice for devices: %w", err)
+	}
 	return nil
 }
 
 func (s *DeviceState) applyConfig(ctx context.Context, config configapi.Interface, claim *resourceapi.ResourceClaim, results []*resourceapi.DeviceRequestAllocationResult) (*DeviceConfigState, error) {
+	var err error
+	var configState DeviceConfigState
+
 	switch castConfig := config.(type) {
 	case *configapi.GpuConfig:
-		return s.applySharingConfig(ctx, castConfig.Sharing, claim, results)
+		configState.GpuConfig = castConfig
+		err = s.applyGpuConfig(ctx, castConfig, claim, results, &configState)
 	case *configapi.MigDeviceConfig:
-		return s.applySharingConfig(ctx, castConfig.Sharing, claim, results)
+		err = s.applySharingConfig(ctx, castConfig.Sharing, claim, results, &configState)
 	case *configapi.ImexChannelConfig:
-		return s.applyImexChannelConfig(ctx, castConfig, claim, results)
+		err = s.applyImexChannelConfig(ctx, castConfig, claim, results, &configState)
 	default:
-		return nil, fmt.Errorf("unknown config type: %T", castConfig)
+		err = fmt.Errorf("unknown config type: %T", castConfig)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &configState, nil
+}
+
+func (s *DeviceState) applyGpuConfig(ctx context.Context, config *configapi.GpuConfig, claim *resourceapi.ResourceClaim, results []*resourceapi.DeviceRequestAllocationResult, configState *DeviceConfigState) error {
+	if config.Sharing != nil {
+		err := s.applySharingConfig(ctx, config.Sharing, claim, results, configState)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Question: We unconditionally call `applySharingConfig` for mig devices. Since we now have the case where `config.Sharing` can be nil, could we either update `applySharingConfig` to ignore nil configs, or introduce a `Disabled` config so that we can set that for cases where we don't expect sharing to be allowed?

Reply:
For MIG devices, `config.Sharing` can never be nil right - MIGs would always support `Sharing` so `Normalize` would initialize that config.